odoo/sql_db.py

```diff
@@ -413,13 +413,26 @@ def __exit__(self, exc_type, exc_value, traceback):
     def savepoint(self):
         """context manager entering in a new savepoint"""
         name = uuid.uuid1().hex
+        env = None
+        for environment in Environment.envs:
+            if environment.cr is self:
+                env = environment
+                break
+        if env:
+            env['base'].flush()
         self.execute('SAVEPOINT "%s"' % name)
         try:
-            yield
+            if env:
+                with env.clear_upon_failure():
```

Comment:
Because of the `flush()` above, `env` should be clear here.
So you could instead add `env.clear()` in the `except` below.